File: perl/hello.pl
#!/usr/bin/env perl
#
# Perl language Hello World! (and then some...) 
use strict;
use warnings;

my @animals = ("camel", "llama", "owl");
my %fruit_color = (apple => "red", banana => "yellow");
my @fruits = keys %fruit_color;
my @colors = values %fruit_color;
my $verinfo = `git describe --abbrev=7 --dirty --always --tags`;

# Obligatory Hello World with some version info
print "Hello Perl World!  Version:  $verinfo\n";

# Show list of command line arguments
print "Command line args: ";
foreach (@ARGV) {
	print "  $_";
}
print "\n";

# Determine number of physical CPU sockets
my $socketcount=`cat /proc/cpuinfo | grep 'physical id' | sort | uniq | wc -l`;
print "CPU socket count: $socketcount\n";

# print out our animals array in reverse-sorted order
print "\nValue of animals array:\n";
foreach (reverse @animals) {
	print 'item = ' .  $_ . "\n";
	if ($_ eq $animals[0]) {
		print "   Condition is TRUE, \"$_\" eq " . "$animals[0]\n";
	} else {
		print "   Condition is FALSE, \"$_\" ne " . "$animals[0]\n";
	}
}

# print our fruit_color hash (key/value pairs)
print "\nValue of fruit_color hash:\n";
foreach (keys %fruit_color) {
	print "$_ = $fruit_color{$_}\n";
}

# Read a non-existent environment variable and supply default value...
my $some_env_var;
if (defined $ENV{'PROBABLY_NOT_THERE'}) {
	$some_env_var = $ENV{'PROBABLY_NOT_THERE'};
} else {
	$some_env_var = "Default value";
}
print "\nPROBABLY_NOT_THERE is set to \'$some_env_var\'\n\n";

# Let's read in the PATH variable from the OS and print it out...
my $path_env_var = $ENV{'PATH'};
print "OS PATH var is set to: \'$path_env_var\'\n\n";

# Let's manipulate PATH variable and then restore it
print "Adding :/test to PATH variable\n";
$ENV{'PATH'} = "$ENV{'PATH'}:/test";
print "OS PATH var is set to: \'$ENV{'PATH'}\'\n\n";
print "Restoring original PATH value.\n";
$ENV{'PATH'} = $path_env_var;
print "OS PATH var is set to: \'$path_env_var\'\n\n";

# Let's create the file test.out and write animals array to it...
print "\nWriting test.out file...\n";
open(my $out, ">", "test.out") or die "Can't create/open test.out! $!\n";
foreach (@animals) {
	print $out "$_\n";
}
close $out or die "$out: $!\n";

# Now open the file we wrote and read it back...
print "\nContents of test.out file:\n";
open (my $in, "<", "test.out") or die "Can't open test.out! $!\n";
my $linenum = 1;
while (<$in>) {
	print "line $linenum: $_";
	$linenum += 1;
}
close $in or die "$in $!\n";

File: python/hello.py
#!/usr/bin/env python
#
# Sample Hello World! python script...
#

import subprocess
import sys
import os

animals = ["camel", "llama", "owl"]
fruit_color = { "apple":"red", "banana":"yellow"}
fruits = fruit_color.keys()
colors = fruit_color.items()
verinfo = subprocess.check_output(["git", "describe", "--abbrev=7",
		"--dirty", "--always", "--tags"])

# Obligatory Hello World with some version info
print("Hello Python World!  Version: {}".format(verinfo))

# Show list of command line arguments
print("Command line args: "),
for i in sys.argv[1:]:
	print(" {}".format(i))
print("")

# Determine number of physical CPU sockets 
if os.name == 'nt':
	pass
else:
	socketcount = subprocess.check_output(["bash","-c",
            "(cat /proc/cpuinfo | grep 'physical id' | sort | uniq | wc -l)"])
	print("CPU socket count: {}".format(socketcount))

# print out our animals array in reverse-sorted order
print("\nValue of animals array:")
for i in sorted(animals, reverse=True):
	print("item = " + i)
	if (i == animals[0]):
		print("   Condition is TRUE, \"" +i+ "\" == " + animals[0]) 
	else:
		print("   Condition is FALSE, \"" +i+ "\" != " + animals[0]) 

# print our fruit_color dictionary (key/value pairs)
print("\nValue of fruit_color hash:")
for fruit,color in sorted(fruit_color.items(), reverse=True):
	print(fruit, " = ", color)

# Read a non-existent environment variable and supply default value...
some_env_var = os.getenv("PROBABLY_NOT_THERE", "Default Value")
print("\nPROBABLY_NOT_THERE is set to '"+some_env_var+"'\n")

# Let's read in the PATH variable from the OS and print it out...
path_env_var = os.environ['PATH']
print("OS PATH var is set to: '"+path_env_var+"'\n")

# Let's manipulate PATH variable and then restore it
print("Adding :/test to PATH variable")
os.environ['PATH'] = os.environ['PATH'] + ":/test"
print("OS PATH var is set to: '"+os.environ['PATH']+"'\n")
print("Restoring original PATH value.")
os.environ['PATH'] = path_env_var
print("OS PATH var is set to: '"+path_env_var+"'\n")


# Let's create the file test.out and write animals array to it...
print("\nWriting test.out file...")
file = open("test.out", "w")
for i in animals:
	file.write(i+"\n")
file.close()

# Now open the file we wrote and read it back...
print("\nContents of test.out file:")
file = open("test.out", "r")
linenum = 1
for line in file:
	print("line "+str(linenum)+": "+line)
	linenum += 1
file.close()

# All done, finish with normal exit code.
sys.exit(0)
